flink-core/src/main/java/org/apache/flink/util/PropertiesUtil.java

```diff
@@ -108,6 +109,27 @@ public static boolean getBoolean(Properties config, String key, boolean defaultV
 		}
 	}
 
+	/**
+	 * Flatten a recursive {@link Properties} to a first level property map.
+	 * In some cases, {KafkaProducer#propsToMap} for example, Properties is used purely as a HashMap
+	 * without considering its default properties.
+	 *
+	 * @param config Properties to be flatten
+	 * @return Properties without defaults; all properties are put in the first-level
+	 */
+	public static Properties flatten(Properties config) {
```

Comment:
> On a second thought, wouldn't it make more sense to provide a correct `propsToMap` implementation? If it's only used in KafkaProducer, then we could fix it there. If not, I'd consider that function more useful than this `flatten`.

this name is still confusing since it is a Properties, not a map.

The flattened properties are actually used in the Kafka client lib, not that easy to fix.

Reply:
> Should be covered with one test case.

sure, will add one later.